src/typechecker/FStar.TypeChecker.TcEffect.fst

```diff
@@ -95,56 +95,1031 @@ let pure_wp_uvar env (t:typ) (reason:string) (r:Range.range) : term * guard_t =
+                             bind_name
+                             (Print.term_to_string bind_t)) r
+    else rest_bs, [] in
+
+  // f binder with sort m_repr ?us
+  let f, guard_f =
+    let repr, g = TcUtil.fresh_effect_repr
+      (Env.push_binders env (a_b::b_b::rest_bs))
+      r
+      m_eff_name
+      m_sig_ts
+      m_repr_ts
+      (U_name u_a)
+      (a_b.binder_bv |> S.bv_to_name) in
+    repr |> S.gen_bv "f" None |> S.mk_binder, g in
+
+  // g binder with sort (x:a -> n_repr ?us)
+  let g, guard_g =
+    let x_a = a_b.binder_bv |> S.bv_to_name |> S.gen_bv "x" None |> S.mk_binder in
+    let repr, g = TcUtil.fresh_effect_repr
+      (Env.push_binders env (a_b::b_b::rest_bs@[x_a]))
+      r
+      n_eff_name
+      n_sig_ts
+      n_repr_ts
+      (U_name u_b)
+      (b_b.binder_bv |> S.bv_to_name) in
+    S.gen_bv "g" None (U.arrow [x_a] (S.mk_Total' repr (Some (new_u_univ ())))) |> S.mk_binder,
+    g in
+
+  // return repr type p_repr ?us
+  let return_repr, guard_return_repr = TcUtil.fresh_effect_repr
+    (Env.push_binders env (a_b::b_b::rest_bs))
+    r
+    p_eff_name
+    p_sig_ts
+    p_repr_ts
+    (U_name u_b)
+    (b_b.binder_bv |> S.bv_to_name) in
+
+  let pure_wp_uvar, g_pure_wp_uvar = pure_wp_uvar
+    (Env.push_binders env (a_b::b_b::rest_bs))
+    return_repr
+    (BU.format1 "implicit for pure_wp in checking bind %s" bind_name)
+    r in
+
+  let k = U.arrow (a_b::b_b::(rest_bs@range_bs@[f; g])) (S.mk_Comp ({
+    comp_univs = [Env.new_u_univ ()];
+    effect_name = PC.effect_PURE_lid;
+    result_typ = return_repr;
+    effect_args = [pure_wp_uvar |> S.as_arg];
+    flags = [] })) in
+
+  let guard_eq =
+    match Rel.teq_nosmt env k bind_t with
+    | None ->
+      raise_error (Errors.Fatal_UnexpectedEffect,
+                   BU.format2 "Unexpected type of %s (%s)\n"
+                     bind_name
+                     (Print.term_to_string bind_t)) r
+    | Some g -> g in
+
+  Rel.force_trivial_guard env (Env.conj_guards [
+    guard_f;
+    guard_g;
+    guard_return_repr;
+    g_pure_wp_uvar;
+    guard_eq]);
+
+  let k = k |> N.remove_uvar_solutions env |> SS.compress in
+
+  let lopt = bind_combinator_kind env m_eff_name n_eff_name p_eff_name
+    m_sig_ts n_sig_ts p_sig_ts
+    m_repr_ts n_repr_ts p_repr_ts
+    bind_us
+    k
+    num_effect_params
+    has_range_binders in
+
+  let kind = 
+    match lopt with
+    | None ->
+      log_ad_hoc_combinator_warning bind_name r;
+      Ad_hoc_combinator
+    | Some l -> Substitutive_combinator l in
 
-(*
- * For all the layered effects combinators, we enforce that their types are
- *   typeable without using subtyping
- *
- * This is to guard against unsoundness creeping in because of using Untyped uvars
- *   when applying these combinators
- *
- * Essentially we want to ensure that uvars are not introduces at a type different than
- *   what they are used at
- *)
-let check_no_subtyping_for_layered_combinator env (t:term) (k:option typ) =
```

Comment:
With proper typechecking of the indexed effect combinator binders, we don't need all this no subtype typechecking and `eqtype_as_type` business.